RFC/src/RFC-0201_TariScript.md

```diff
@@ -0,0 +1,410 @@
+    Commitment(PedersenCommitment),
+    PublicKey(RistrettoPublicKey),
+    Signature(RistrettoSchnorr),
+}
+```
+
+## Extensions
+
+Tari script places restrictions on _who_ can spend UTXOs. It will also be useful for Tari digital asset applications to
+restrict _how_ or _where_ UTXOs may be spent in some cases. The general term for these sorts of restrictions are termed
+_covenants_. The [Handshake white paper] has a fairly good description of how covenants work.
+
+It is beyond the scope of this RFC, but it's anticipated that Tari Script would play a key role in the introduction of
+generalised covenant support into Tari.
+
+## Applications
+
+### One-sided transactions
+
+One-sided transactions are Mimblewimble payments that do not require the receiver to interact in the transaction
+process. [LIP-004] describes how this will be implemented in Litecoin's Mimblewimble implementation. The main thrust is
+that the sender uses Diffie-Hellman exchange to generate a shared private key that is used as the receiver's blinding
+factor.
+
+To prevent the sender from spending the coins (since both parties now know the spending key), there is an additional
+balance that is performed on the block and transaction that requires the spender to know the receiver's private key.
+
+To implement one-sided payments in Tari, we propose using Diffie-Hellman exchange in conjunction with Tari Script to
+achieve the same thing.
+
+In particular, if Alice is sending some Tari to Bob, she generates a shared private key as follows:
+
+$$ k_s = \mathrm{H}(k_a P_b) $$
+
+where \\( P_b \\) is Bob's Tari node address, or any other public key that Bob has shared with Alice. Alice can generate
+an ephemeral public-private keypair, \\( P_a = k_a\cdot G \\) for this transaction.
+
+Alice then locks the output with the following script:
+
+```text
+Dup PushPubkey(P_B) EqualVerify CheckSig Add
+```
+
+where `P_B` is Bob's public key. As one can see, this Tari script is very similar to Bitcoin script.  
+The interpretation of this script is, "Given a Public key, and a signature of this
+script, the public key must be equal to the one in the locking script, and the signature must be valid using the same
+public key".
+
+This is in effect the same as Bitcoin's P2PK script. To increase privacy, Alice could also lock the UTXO with a P2PKH
+script:
+
+```text
+Dup HashBlake256 PushHash(HB) EqualVerify CheckSig Add
+```
+
+where `HB` is the hash of Bob's public key.
+
+In either case, only someone with the knowledge of Bob's private key can generate a valid signature, so Alice will not
+be able to unlock the UTXO to spend it.
+
+Since the script is committed to and cannot be cut-through, only Bob will be able to spend this UTXO unless someone is
+able to discover the private key from the public key information (the discrete log assumption), or if the majority of
+miners collude to not honour the consensus rules governing the successful evaluation of the script (the 51% assumption).
```

Comment:
The only problem is you would need an archival node to prove this. 

I think to increase trust we might need to put some limit on how long the script + result is stored. 
Take for example we have a case where a node receives a string of 10 valid headers.
 Current MW states that f you receive that 10 headers + state, you can 100% verify that nothing bad happend there. If we include tari script we cannot say with 100% surety that  nothing bad happend. This is especially important if we want to have one sided tx's where we more rely on the security of the script. 
In bitcoin you dont have pruned node state so all nodes will have to provide you those 10 blocks. 

In practice I think our current nodes won't accept that and will just continue mining on the current chain. But for network integrity and locking this down,  I think we need to add a consensus to node that forces a limit on what the minimum pruning horizon is. 

On coins like grin, you can run with a pruning horizon of 1, and altho you will have to reorg a lot, you can alway be 100% sure that nothing bad happend. If you do that here with tari_script you cannot say that.  If we enforce a limit of say coinbase_lockheight blocks or so pruning_horizon you know with a fair amount of security that you will see that tx and its script. It is checked and  was available to be checked for a while.

Reply:
Yes, that's a valid point.

We could require keeping scripts until the pruning horizon, which is a few thousand blocks. So an attack is still possible if there's a very, very deep re-org.

The best way to block that attack is to re-spend coins to yourself with a default script (aka normal Tari tx) where possible. This is simple for one-sided transactions, but easier said than done, for say peg-ins.

Reply:
I'm also thinking that we don't, as a rule have scripts where there is no blinding factor, or where the blinding factor is known, but use DH shared secrets in nearly all cases. This removes the honeypot where miners can say "if only the script was gone, we could spend this coin"

Reply:
That would help. Its going to help make it harder to break. Its going to take some colluding to get the sender + miner in on it.